File: trading-backend/src/core/ports/watchlist_repository.py
from abc import ABC, abstractmethod

from src.core.entities.watchlist import Watchlist


class WatchlistRepositoryPort(ABC):
    """Port for persisting user watchlists and their asset associations."""

    @abstractmethod
    async def save(self, watchlist: Watchlist) -> Watchlist: ...

    @abstractmethod
    async def get_by_id(self, watchlist_id: str) -> Watchlist | None: ...

    @abstractmethod
    async def get_by_user(self, user_id: str) -> list[Watchlist]: ...

    @abstractmethod
    async def delete(self, watchlist_id: str) -> None: ...

    @abstractmethod
    async def add_asset(self, watchlist_id: str, asset_id: str) -> None: ...

    @abstractmethod
    async def remove_asset(self, watchlist_id: str, asset_id: str) -> None: ...

File: trading-frontend/src/features/watchlists/domain/repositories/WatchlistRepository.ts
import { Watchlist } from '@/features/watchlists/domain/entities/Watchlist';

export interface WatchlistRepository {
  getAll(): Promise<Watchlist[]>;
  getById(id: string): Promise<Watchlist>;
  create(name: string): Promise<Watchlist>;
  delete(id: string): Promise<void>;
  addAsset(watchlistId: string, assetId: string): Promise<void>;
  removeAsset(watchlistId: string, assetId: string): Promise<void>;
}
